app/presenters/configuration_status_presenter.rb

```diff
@@ -0,0 +1,254 @@
+      .or(always_show_configuration_status)
+  end
+
+  def always_show_configuration_status
+    settings_table
+      .where(settings_table[:name].eq('always_show_configuration_status')
+        .and(settings_table[:value].eq('--- true')
+          .or(settings_table[:value].eq(nil).and(settings_table[:default].eq('--- true')))))
+      .exists
+  end
+
+  # 0000 0000000000 0000000000 0000000000 0000000000 0000000000 0000000000
+  # ----     [6]        [5]        [4]        [3]        [2]        [1]
+  # [1] - applied
+  # [2] - restarted
+  # [3] - failed
+  # [4] - failed_restarts
+  # [5] - skipped
+  # [6] - pending
+
+  # WHEN pending THEN pending
+  # WHEN failed OR failed_restarts THEN error
+  # WHEN applied OR restarted THEN active
+  # ELSE no_changes
+
+  def computed_status
+    arel_case
+      .when(alerts_disabled).then(status_class::ALERTS_DISABLED)
+      .when(no_reports).then(status_class::NO_REPORTS)
+      .when(out_of_sync).then(status_class::OUT_OF_SYNC)
+      .when(pending).then(status_class::PENDING)
+      .when(failed_or_failed_restarts).then(status_class::ERROR)
+      .when(applied_or_restarted).then(status_class::ACTIVE)
+      .else(status_class::NO_CHANGES)
+      .as('computed_status')
+  end
+
+  def alerts_disabled
+    hosts_table[:enabled].eq(false)
+  end
+
+  def any_reports
+    reports_table
+      .where(reports_table[:type].eq('ConfigReport')
+        .and(reports_table[:host_id].eq(host_status_table[:host_id])))
+      .exists
+  end
+
+  def no_reports
+    any_reports.not
+  end
+
+  def pending
+    status = host_status_table[:status] & 0b1111111111_0000000000_0000000000_0000000000_0000000000_0000000000
+
+    status.gteq(0b1_0000000000_0000000000_0000000000_0000000000_0000000000)
+      .and(status.lteq(0b1111111111_0000000000_0000000000_0000000000_0000000000_0000000000))
+  end
+
+  def failed_or_failed_restarts
+    status = host_status_table[:status] & 0b1111111111_1111111111_0000000000_0000000000
+
+    status.gteq(0b1_0000000000_0000000000)
+      .and(status.lteq(0b1111111111_1111111111_0000000000_0000000000))
+  end
+
+  def applied_or_restarted
+    status = host_status_table[:status] & 0b1111111111_1111111111
+
+    status.gteq(0b1)
+      .and(status.lteq(0b1111111111_1111111111))
+  end
+
+  def out_of_sync
+    out_of_sync_disabled.not
+      .and(host_status_table[:id].in(out_of_sync_ids))
+  end
+
+  def out_of_sync_disabled
+    settings_table
+      .where(settings_table[:name].in(
+        reports_table
+          .where(reports_table[:type].eq('ConfigReport')
+          .and(reports_table[:host_id].eq(host_status_table[:host_id])))
+          .order(reports_table[:reported_at].desc)
+          .take(1).project(
+            arel_concat(reports_table[:origin].lower, arel_quoted('_out_of_sync_disabled'))
+          ))
+        .and(settings_table[:value].eq('--- true'))
+        .or(settings_table[:value].eq(nil).and(settings_table[:default].eq('--- true'))))
+      .exists
```

Comment:
:exploding_head:
Can't we have an additional query `ConfigReport.all.pluck(:origin)` get a setting for each of them and have an sql case instead of this? :thinking: